contracts/services/osmosis-gamm-lper/src/balancer.rs

```diff
@@ -0,0 +1,219 @@
+}
+
+pub fn provide_double_sided_liquidity(
+    deps: DepsMut,
+    cfg: Config,
+    expected_spot_price: Option<DecimalRange>,
+) -> Result<Response, ServiceError> {
+    // first we assert the input account balances
+    let bal_asset_1 = deps
+        .querier
+        .query_balance(&cfg.input_addr, &cfg.lp_config.pool_asset_1)?;
+    let bal_asset_2 = deps
+        .querier
+        .query_balance(&cfg.input_addr, &cfg.lp_config.pool_asset_2)?;
+
+    let pm_querier = PoolmanagerQuerier::new(&deps.querier);
+
+    let pool_ratio = pm_querier.query_spot_price(&cfg.lp_config)?;
+    let pool = pm_querier.get_pool_response(&cfg.lp_config)?;
+
+    // assert the spot price to be within our expectations,
+    // if expectations are set.
+    if let Some(acceptable_spot_price_range) = expected_spot_price {
+        acceptable_spot_price_range.contains(pool_ratio)?;
+    }
+
+    let provision_coins = calculate_provision_amounts(bal_asset_1, bal_asset_2, pool_ratio)?;
+
+    let share_out_amt =
+        calculate_share_out_amt_no_swap(&deps, cfg.lp_config.pool_id, provision_coins.clone())?;
+
+    let liquidity_provision_msg: CosmosMsg = get_provide_liquidity_msg(
+        cfg.input_addr.as_str(),
+        cfg.lp_config.pool_id,
+        provision_coins,
+        share_out_amt.to_string(),
+    )?;
+
+    let transfer_lp_tokens_msg = BankMsg::Send {
+        to_address: cfg.output_addr.to_string(),
+        amount: coins(
+            share_out_amt.u128(),
+            pool.total_shares
+                .ok_or_else(|| StdError::generic_err("failed to get shares"))?
+                .denom,
+        ),
+    };
+
+    let delegated_msgs = execute_on_behalf_of(
+        vec![liquidity_provision_msg, transfer_lp_tokens_msg.into()],
+        &cfg.input_addr.clone(),
+    )?;
+
+    Ok(Response::default().add_message(delegated_msgs))
+}
+
+pub trait ValenceLiquidPooler {
+    fn query_spot_price(&self, lp_config: &LiquidityProviderConfig) -> StdResult<Decimal>;
+    fn get_pool_response(&self, lp_config: &LiquidityProviderConfig) -> StdResult<Pool>;
+}
+
+impl ValenceLiquidPooler for PoolmanagerQuerier<'_, Empty> {
+    fn query_spot_price(&self, lp_config: &LiquidityProviderConfig) -> StdResult<Decimal> {
+        let spot_price_response = self.spot_price(
+            lp_config.pool_id,
+            lp_config.pool_asset_1.to_string(),
+            lp_config.pool_asset_2.to_string(),
+        )?;
+
+        let pool_ratio = Decimal::from_str(&spot_price_response.spot_price)?;
+
+        Ok(pool_ratio)
+    }
+
+    fn get_pool_response(&self, lp_config: &LiquidityProviderConfig) -> StdResult<Pool> {
+        let pool_response = self.pool(lp_config.pool_id)?;
+
+        let pool: Pool = pool_response
+            .pool
+            .ok_or_else(|| StdError::generic_err("failed to get pool"))?
+            .try_into()
+            .map_err(|_| StdError::generic_err("failed to decode proto"))?;
+
+        Ok(pool)
+    }
+}
+
```

Comment:
should prob be renamed right? You are querying the pool ratio and the pool

Reply:
renamed the `get_pool_response` to `query_pool`. as for the pool ratio, given that it's the balancer pool, the ratio is the spot price so felt like it describes the function a little better. do you think we should stick to pool ratio?